### DISCUSSION TITLE
dll files not found after adding Path environment variable

### DISCUSSION THREAD

**Discussion by justincase82** [thirtythreeforty/neolink#167] [General]
I'm running Windows 10 Pro 20H2 and I have the release-windows-2019 package. 

I've followed the install instructions and downloaded and installed GStreamer MSI package. 

%GSTREAMER_1_0_ROOT_X86_64%\bin has been added to the system "path" variable but it keeps complaining about missing .dll files. 

I've tried a few different GStreamer packages with the same results.

https://gstreamer.freedesktop.org/data/pkg/windows/1.14.5/gstreamer-1.0-x86_64-1.14.5.msi
https://gstreamer.freedesktop.org/data/pkg/windows/1.19.1/mingw/gstreamer-1.0-mingw-x86_64-1.19.1.msi
https://gstreamer.freedesktop.org/data/pkg/windows/1.2.3/gstreamer-1.0-x86_64-1.2.3.msi

powershell dir env: shows 

GSTREAMER_1_0_ROOT_X86_64      C:\gstreamer\1.0\x86_64\

Path                           C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;%GSTREAMER_1_0_ROOT_X86_64%\bin;C:\Users\cameras\AppData\Local\Microsoft\WindowsApps;

I'm launching via CMD "c:\neolink\neolink --config config.toml"

What am I doing wrong here?

Thanks for the help.

-Justin

**Comment by twistedddx** [comment]
Use the msvc builds not the mingw.
Do complete installs to ensure you install all the required components.
Install both the runtime and development installers also.
https://gstreamer.freedesktop.org/data/pkg/windows/1.18.4/msvc/gstreamer-1.0-msvc-x86_64-1.18.4.msi
https://gstreamer.freedesktop.org/data/pkg/windows/1.18.4/msvc/gstreamer-1.0-devel-msvc-x86_64-1.18.4.msi

**Reply by justincase82** [reply]
I had the same issues, but then realized I was not running the script from an admin prompt. Too many beers last night LOL

Neolink now launches correctly.

Thanks for the reply

**Comment by stegl** [comment]
hello, 

well i have the same problem and cant fix it ... i the twistedddx recommendation. 
i have both neolink and gstreamer on C
tryed running cmd as administrator

if you need some  more information please tell me what you need and ill post.
 
![1](https://user-images.githubusercontent.com/65857447/122656392-72e7fc80-d15a-11eb-9005-2c0a7384865b.png)
![2](https://user-images.githubusercontent.com/65857447/122656394-74b1c000-d15a-11eb-947d-775996b4be11.png)
![3](https://user-images.githubusercontent.com/65857447/122656395-754a5680-d15a-11eb-9538-940973bf03da.png)
![4](https://user-images.githubusercontent.com/65857447/122656397-75e2ed00-d15a-11eb-8f54-afd8c73f8e5f.png)
 
thanks in advance

**Reply by QuantumEntangledAndy** [reply]
@stegl your error suggests you havent installed gstreamer. Please ensure you install the msvc version of gstreamer. 

During the MSI install you should also select the full installation so that it installs the rtsp plugins. 

If you think you have please check the `C:\gstreamer\1.0\x86_64\` directory and ensure all DLLs are listed in there.

**Reply by stegl** [reply]
yes i did install msvc version and selected full install... 

i do have installed gtstreamer but its in different folder..?
![5](https://user-images.githubusercontent.com/65857447/122675191-9ce00400-d1d8-11eb-9f7e-e6f843c671c0.png)

in C:\gstreamer\1.0\msvc_x86_64\bin folder is 243 files... if this tells something 

i installed
gstreamer-1.0-msvc-x86_64-1.18.4.msi and 
gstreamer-1.0-devel-msvc-x86_64-1.18.4.msi

thank you for your answer

**Reply by QuantumEntangledAndy** [reply]
Then the next step is to check your `PATH` and your `GSTREAMER_1_0_ROOT_X86_64` environmental variables. These should be set be the gstreamer installer but perhaps that part didn't work for some reason. Do you know how to check these variables?

**Reply by stegl** [reply]
![6](https://user-images.githubusercontent.com/65857447/122685734-8607d480-d20d-11eb-81e7-52f5034b80a9.png)

i did follow the instructions on github.. 


didnt understand this part "Download and extract a prebuilt binary from continuous integration " but then i found the download link.

**Reply by twistedddx** [reply]
Your variable has MSVC in the name but the path doesn't have the MSVC.

Gstreamer installer is stupid. Change the variable name to GSTREAMER_1_0_ROOT_X86_64

**Reply by QuantumEntangledAndy** [reply]
As @twistedddx pointed out your variables names are not matching. 

![image](https://user-images.githubusercontent.com/13386481/122695118-b4b39880-d269-11eb-92c0-cc046ddd7d74.jpeg)

**Reply by stegl** [reply]
i tried fixing it both ways.... but it doesn't work any way. 
![7](https://user-images.githubusercontent.com/65857447/122728179-2397ee00-d278-11eb-833d-c69406166c22.png)

![8](https://user-images.githubusercontent.com/65857447/122728304-488c6100-d278-11eb-85e8-7d05c445319f.png)

i did take a look in /bin folder and there are all dll s needed

i treyed to add and remove /bin but it all the same result
%GSTREAMER_1_0_ROOT_X86_64%\bin
%GSTREAMER_1_0_ROOT_X86_64%
C:\gstreamer\1.0\msvc_x86_64\bin
C:\gstreamer\1.0\msvc_x86_64

**Reply by QuantumEntangledAndy** [reply]
I think you are not setting it up correctly....

1. create a variable
   `GSTREAMER_ROOT_X86_64` with the contents `C:\gstreamer\1.0\msvc_x86_64\`

2. Modify the `PATH` to contain
   `%GSTREAMER_ROOT_X86_64%\bin`

Remember that `PATH` is a `;` separated list of locations

**Reply by stegl** [reply]
like this? or do i remove user variables? right now is still not working 
![10](https://user-images.githubusercontent.com/65857447/122731838-f51c1200-d27b-11eb-9e99-65c1278213e4.png)

**Reply by QuantumEntangledAndy** [reply]
No you have not set up `PATH` correctly. Please read the instructions carefully.

In the PATH (both of them) you need to include the `%` around the  `GSTREAMER_ROOT_X86_64` and after the `%` put in `\bin` such that it read

`%GSTREAMER_ROOT_X86_64%\bin`

**Reply by stegl** [reply]
still not working
![11](https://user-images.githubusercontent.com/65857447/122741524-54caeb00-d285-11eb-8022-655e103a81f9.png)

on user variables makes it \bin\bin\bin at the end so i deleted it (pic below)
![12](https://user-images.githubusercontent.com/65857447/122741938-c30fad80-d285-11eb-93db-eeeef4ed2159.png)

**Reply by twistedddx** [reply]
stegl, the environment variables are very important to your system and you should be very careful when altering them to ensure you are making correct changes.

Current gstreamer installers create the user variable:
name: GSTREAMER_1_0_ROOT_MSVC_X86_64
value: C:\gstreamer\1.0\msvc_x86_64\

Note: no bin!

I suggested you change the variable name from:
GSTREAMER_1_0_ROOT_MSVC_X86_64
to:
GSTREAMER_1_0_ROOT_X86_64


The other way to attack it would be to alter the Path to include %GSTREAMER_1_0_ROOT_MSVC_X86_64%\bin
But altering Path is more dangerous as you could remove needed paths by accident.

It looks like you already entirely deleted your user variable for Path which is pretty bad for your Windows installation in general.
It is possible you need to do a system restore to before you did that to undo that change.


Changes to Environment variables do not apply until after a restart or you can force a refresh by killing explorer.exe and relaunching explorer.exe.
I would suggest you do the restart option as it ensures the change will apply and prevents doing something wrong.

The other other way of attacking this is to put neolink.exe in the C:\gstreamer\1.0\msvc_x86_64\bin folder.
That would mean you can ignore Environment variables.

**Reply by QuantumEntangledAndy** [reply]
You also failed to follow my instructions again... ToT 

1. create a variable
    `GSTREAMER_ROOT_X86_64` with the contents `C:\gstreamer\1.0\msvc_x86_64\`

2. Modify the PATH to contain
    %GSTREAMER_ROOT_X86_64%\bin

DO NOT set `GSTREAMER_ROOT_X86_64` to `%GSTREAMER_ROOT_X86_64%/bin` that would just be cyclic since you are setting `GSTREAMER_ROOT_X86_64` to depend on `GSTREAMER_ROOT_X86_64` which weird.

**Reply by stegl** [reply]
@twistedddx its okay i didnt delete anything or i have it saved. i tryed as you said but it still doesnt work. i did full restart. 
im remotely connected to machine so it takes alot of time to restart :P 

@QuantumEntangledAndy 
sorry. 

btw from this:

"2. If you are using Windows, add the following to your PATH environment variable:

%GSTREAMER_1_0_ROOT_X86_64%\bin"

i would say you just put this "%GSTREAMER_1_0_ROOT_X86_64%\bin" in System variables PATH. That **System** variables just means so much. But thats just me. 



i will put neolink in bin folder.. i dont mind it there if it will work. btw do i put all files or just neolink.exe, neolink.d and config?

**Reply by QuantumEntangledAndy** [reply]
No worries your new to this and we all start somewhere.


You only need to include the `neolink` not the `neolink.d`.

**Reply by stegl** [reply]
well sorry again...
i put neolink id bin folder and it looks like its working but i cant find streams.... 
![brrr](https://user-images.githubusercontent.com/65857447/123008707-abd2dc00-d3bb-11eb-8ed9-51a480edefc9.png)

are that errors on top actually error? 

it should look like this right?
"rtsp://username:pass@192.168.1.226/podstresk"

username:pass the one i set in config?

**Reply by twistedddx** [reply]
You need the port number which by default is 8554
Eg Neolink instructions say the url is:
rtsp://127.0.0.1:8554/your_camera_name

**Reply by stegl** [reply]
sorry.. thank you verry mutch all. it works.

**Comment by Ayumon3** [comment]
I'm also having this problem. I have installed GStreamer MSVC x86-64 version 1.20.0 runtime using a complete install. But it looks like a slightly different issue to above.

After checking all the thread replies, I have verified that my path is set correctly. The suggested system path value of "%GSTREAMER_1_0_ROOT_X86_64%\bin" didn't work because the user variable set by the install was slightly different. My solution to this was different - I simply set the system path to be explicity rather than based on the variable. Having read this thread, I've returned to the variable method, and fixed the user variable. Either way, my problem remains. 

My user variables updated according to thread recommendations.
![image](https://user-images.githubusercontent.com/52187979/154400933-d887c069-30c5-4257-98bb-7b8a53f73350.png)
My system path variables are referencing my user variables. The path ends in /bin so that the binaries can be called.
![image](https://user-images.githubusercontent.com/52187979/154401025-6fc10a47-d4e0-4682-ae67-6ac4aafa0cd6.png)

I can run other binaries from that path, such as gsettings.exe
![image](https://user-images.githubusercontent.com/52187979/154397866-82f726eb-277e-44c7-92a1-f8191f814b73.png)

I have verified that I'm using administrative command prompt.
![image](https://user-images.githubusercontent.com/52187979/154397635-696b71b1-beab-4e86-a99f-3a2714a5bb0b.png)

I have verified that the DLLs listed in the warnings are present in the path described in the error.
![image](https://user-images.githubusercontent.com/52187979/154397806-f295060b-57c5-4981-a26b-393fbb113b4d.png)

I have tried installing the developer files as well, but that didn't help. The reality is that the DLLs the program can't access are there, but something is stopped the program from accessing them.

Any ideas?

**Reply by Ayumon3** [reply]
Despite the above error, I can get a video feed from the camera in BlueIris. This is awesome. I'd love to buy the dev a coffee.

I wonder what the DLL does. with a name like "gstwavpack.dll" it sounds like it could be something to do with sound. Edit: I can confirm that the audio feed isnt working.

Below are my BlueIris settings
![image](https://user-images.githubusercontent.com/52187979/154401652-048efa39-befd-48cf-9757-e8e31d5d19a1.png)
